File: Java/SelectionSort.java
import java.util.Scanner;

class SelectionSort {
	void sort(int arr[]) {
		int n = arr.length;

		// One by one move boundary of unsorted subarray
		for (int i = 0; i < n - 1; i++) {
			// Find the minimum element in unsorted array
			int min_idx = i;
			for (int j = i + 1; j < n; j++)
				if (arr[j] < arr[min_idx])
					min_idx = j;

			// Swap the found minimum element with the first
			// element
			int temp = arr[min_idx];
			arr[min_idx] = arr[i];
			arr[i] = temp;
		}
	}

	// Prints the array
	void printArray(int arr[]) {
		int n = arr.length;
		for (int i = 0; i < n; ++i)
			System.out.print(arr[i] + " ");
		System.out.println();
	}

	// Driver code to test above
	public static void main(String args[]) {
		SelectionSort ob = new SelectionSort();
		Scanner sc = new Scanner(System.in);

		// Input the size of the array
		System.out.println("Enter the size of array: ");
		int size = sc.nextInt();

		// Initialize the array with given size
		int arr[] = new int[size];

		// Input the array elements
		System.out.println("Enter array elements");
		for (int i = 0; i < size; i++)
			arr[i] = sc.nextInt();
		ob.sort(arr);
		System.out.println("Sorted array");
		ob.printArray(arr);
	}
}
/*
// Java program for Implementation of Selection Sort
Input: Size of array, array of integers
Output: Sorted array in ascending order
----------------------------------------
Sample Input:
4
4 3 2 1
Output:
Sorted array
1 2 3 4
----------------------------------------
Time Complexity: O(n^2)
Space Complexity: O(1)
*/

File: Kotlin/SelectionSort.kt
fun main(args: Array<String>) {
    //Input the array in the format num1 num2 num3 num4 etc
    print("Enter the elements for the Array: ")
    var inputArray = readLine()!!.trim().split(" ").map { it -> it.toInt() }.toIntArray()

    var sortedArray = selectionSort(inputArray)
    println(sortedArray.contentToString())
}

fun selectionSort(arr: IntArray): IntArray {
    for (i in arr.lastIndex downTo 0) {
        var max = arr[0]
        var maxIndex = 0
        for (j in 0..i) {
            if (arr[maxIndex]<arr[j])
                maxIndex = j
        }
        val temp = arr[maxIndex]
        arr[maxIndex] = arr[i]
        arr[i] = temp
    }
    return arr
}

//Sample
//Input:   Enter the elements for the Array: 64 23 -12 -5 0 2 89 20 100 32 650 -230 130
//Output:   [-230, -12, -5, 0, 2, 20, 23, 32, 64, 89, 100, 130, 650]

//Time Complexity:  Worst Case: O(n^2)   Best Case: O(n^2)
//Space Complexity: O(1)
